modules/table/src/main/java/org/apache/ignite/internal/table/distributed/StorageUpdateHandler.java

```diff
@@ -157,51 +158,77 @@ public void handleUpdateAll(
-                // Sort IDs to prevent deadlock. Natural UUID order matches RowId order within the same partition.
-                SortedMap<UUID, TimedBinaryRow> sortedRowsToUpdateMap = new TreeMap<>(rowsToUpdate);
-
-                for (Map.Entry<UUID, TimedBinaryRow> entry : sortedRowsToUpdateMap.entrySet()) {
-                    RowId rowId = new RowId(partitionId, entry.getKey());
-                    BinaryRow row = entry.getValue() == null ? null : entry.getValue().binaryRow();
-
-                    locker.lock(rowId);
+        if (nullOrEmpty(rowsToUpdate)) {
+            return;
+        }
+        long batchSize = getBatchSize();
+        BiPredicate<List<Entry<UUID, TimedBinaryRow>>, Entry<UUID, TimedBinaryRow>> isFullByRowCount = (batch, row) ->
+                batch.size() + 1 > batchSize;
+        List<List<Map.Entry<UUID, TimedBinaryRow>>> batches = getTimedBinaryRowBatches(rowsToUpdate, isFullByRowCount);
+        batches.forEach((batch) ->
+                storage.runConsistently(locker -> {
+                    int commitTblId = commitPartitionId.tableId();
+                    int commitPartId = commitPartitionId.partitionId();
+
+                    List<RowId> rowIds = new ArrayList<>();
+                    for (Map.Entry<UUID, TimedBinaryRow> entry : batch) {
+                        RowId rowId = new RowId(partitionId, entry.getKey());
+                        BinaryRow row = entry.getValue() == null ? null : entry.getValue().binaryRow();
+
+                        locker.lock(rowId);
+
+                        performStorageCleanupIfNeeded(txId, rowId, entry.getValue() == null ? null : entry.getValue().commitTimestamp());
+
+                        if (commitTs != null) {
+                            storage.addWriteCommitted(rowId, row, commitTs);
+                        } else {
+                            BinaryRow oldRow = storage.addWrite(rowId, row, txId, commitTblId, commitPartId);
+
+                            if (oldRow != null) {
+                                assert commitTs == null : String.format("Expecting explicit txn: [txId=%s]", txId);
+                                // Previous uncommitted row should be removed from indexes.
+                                tryRemovePreviousWritesIndex(rowId, oldRow);
+                            }
+                        }
 
-                    performStorageCleanupIfNeeded(txId, rowId, entry.getValue() == null ? null : entry.getValue().commitTimestamp());
+                        rowIds.add(rowId);
+                        indexUpdateHandler.addToIndexes(row, rowId);
+                    }
 
-                    if (commitTs != null) {
-                        storage.addWriteCommitted(rowId, row, commitTs);
-                    } else {
-                        BinaryRow oldRow = storage.addWrite(rowId, row, txId, commitTblId, commitPartId);
+                    if (trackWriteIntent) {
+                        pendingRows.addPendingRowIds(txId, rowIds);
```

Comment:
Before your change, `rowIds` contained all rowIds of this operation; now it only contains IDs related to one batch. I would ask the guys is this is ok that the pending IDs collection is updated in chunks.